### ISSUE TITLE
Diagnose and repair the Elements deployment for cisdr201w

### ISSUE THREAD

**Issue by jrgriffiniii** [pulibrary/dspace-development#558]
The following error message is currently being rendered for the production environment (cisdr201w):

![image.png](https://images.zenhubusercontent.com/5b5b6aed981c3204c9ed06c8/ae0cbb02-ab87-4b3e-8bfe-eaf1d7d5da25)

**Comment by jrgriffiniii** [comment]
This is not related to any X.509 certificate issues:

![image.png](https://images.zenhubusercontent.com/5b5b6aed981c3204c9ed06c8/d2371b33-7f86-400d-88b2-b6c6243e43db)

**Comment by jrgriffiniii** [comment]
No detailed information is provided in the logs for the application when a restart is attempted:

```
#Version: 1.0
#Date: 2021-12-08 14:47:16
#Fields: date time s-ip cs-method cs-uri-stem cs-uri-query s-port cs-username c-ip cs(User-Agent) cs(Referer) sc-status sc-substatus sc-win32-status time-taken
2021-12-08 14:47:16 ***.180.220.210 GET / - 443 - 172.20.208.173 Mozilla/5.0+(Macintosh;+Intel+Mac+OS+X+10_15_7)+AppleWebKit/537.36+(KHTML,+like+Gecko)+Chrome/95.0.4638.69+Safari/537.36 - 302 0 0 2950
2021-12-08 14:47:16 ***.180.220.210 GET /login.html - 443 - 172.20.208.173 Mozilla/5.0+(Macintosh;+Intel+Mac+OS+X+10_15_7)+AppleWebKit/537.36+(KHTML,+like+Gecko)+Chrome/95.0.4638.69+Safari/537.36 - 500 0 0 318
2021-12-08 14:47:16 ***.180.220.210 GET /images/branding/ - 443 - 172.20.208.173 Mozilla/5.0+(Macintosh;+Intel+Mac+OS+X+10_15_7)+AppleWebKit/537.36+(KHTML,+like+Gecko)+Chrome/95.0.4638.69+Safari/537.36 https://oaworkflow.princeton.edu/login.html 403 14 0 32
```

**Comment by jrgriffiniii** [comment]
The following is an error logged for the Elements application:

![image.png](https://images.zenhubusercontent.com/5b5b6aed981c3204c9ed06c8/c25caaee-f189-483d-9336-ecc6861ab5a9)

**Comment by jrgriffiniii** [comment]
The SN@P ticket is currently awaiting a response at https://princeton.service-now.com/service?id=ticket&table=incident&sys_id=40d6fb65879809549cd197183cbb3501&view=sp

**Comment by jrgriffiniii** [comment]
After further trying to reproduce and diagnose this error, the following has been achieved:

- The Elements scheduler service has been manually restarted (it appears to not have been running)
- A more detailed error message has been located

The aforementioned error mentioned is the following:

```
Log entry ID: 4aabbe3d-72ae-475a-961e-bdef9bc3b70e
Logged at: 11:43:41.292 08/12/21

Level 1:
-------
Exception type: Symplectic.Elements.DataEntities.WebsiteServerErrorException
Message: The website encountered a problem while handling a webpage request.
Exception not yet thrown.

Level 2:
-------
Exception type: System.Collections.Generic.KeyNotFoundException
Message: The given key was not present in the dictionary.
Stack trace:
at System.Collections.Generic.Dictionary<TKey, TValue>.get_Item(TKey)
at Symplectic.Elements.DataEntities.ObjectModel.UserCache..ctor(Int64, RMBasicUserData, Dictionary<Int32, RMUserMessageOptions>, List<RMBasicUserData>, List<RMDelegate>, Dictionary<Int32, Group>, Dictionary<Int32, HerdcGroup>, List<RMUserRoleMapEntry>, List<RMGroupRoleMapEntry>, ICache) in elements\DataEntities\ObjectModel\RightsModel.cs:line 1158, char 24
at Symplectic.Elements.DataAccess.ElementsDatabase.GetUserCache() in elements\DataAccess\ElementsDatabase.cs:line 2115, char 5
at Symplectic.Elements.DataAccess.CacheManager.get_UserCache() in elements\DataAccess\CacheManager.cs:line 95, char 7
at Symplectic.Elements.DataEntities.Caching.Cache.get_TheUserCache() in elements\DataEntities\Caching\Cache.cs:line 218, char 5
at Symplectic.Elements.Website.Application.UISerialise.StandardConfiguration(RMUser) in elements\Website.Application\UISerialise.cs:line 53, char 4
at Symplectic.Elements.Website.Application.BaseHandlers.XmlOutputHandler<T>.SerialiseStandard(SafeXmlTextWriter, Boolean) in elements\Website.Application\BaseHandlers\XmlOutputHandler.cs:line 274, char 4
at Symplectic.Elements.Website.Application.Handlers.LoginPage.Respond(LoginProgress) in elements\Website.Application\Handlers\Login.cs:line 532, char 4
at Symplectic.Elements.Website.Application.Handlers.LoginPage.WriteOutputData() in elements\Website.Application\Handlers\Login.cs:line 328, char 5
at Symplectic.Elements.Website.Application.BaseHandlers.XmlOutputHandler<T>.InnerProcessRequest() in elements\Website.Application\BaseHandlers\XmlOutputHandler.cs:line 124, char 5



2021-12-08 11:43:41: System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Symplectic.Elements.DataEntities.ObjectModel.UserCache..ctor(Int64 cacheID, RMBasicUserData systemUser, Dictionary`2 messageOptions, List`1 allUsers, List`1 delegates, Dictionary`2 groups, Dictionary`2 herdcGroups, List`1 userRoles, List`1 groupRoles, ICache cache) in C:\BuildAgent\work\2f923cdb0b81569d\elements\DataEntities\ObjectModel\RightsModel.cs:line 1158
   at Symplectic.Elements.DataAccess.ElementsDatabase.GetUserCache() in C:\BuildAgent\work\2f923cdb0b81569d\elements\DataAccess\ElementsDatabase.cs:line 2115
   at Symplectic.Elements.DataAccess.CacheManager.get_UserCache() in C:\BuildAgent\work\2f923cdb0b81569d\elements\DataAccess\CacheManager.cs:line 95
   at Symplectic.Elements.DataEntities.Caching.Cache.get_TheUserCache() in C:\BuildAgent\work\2f923cdb0b81569d\elements\DataEntities\Caching\Cache.cs:line 218
   at Symplectic.Elements.Website.Application.UISerialise.StandardConfiguration(RMUser impersonatedUser) in C:\BuildAgent\work\2f923cdb0b81569d\elements\Website.Application\UISerialise.cs:line 53
   at Symplectic.Elements.Website.Application.BaseHandlers.XmlOutputHandler`1.SerialiseStandard(SafeXmlTextWriter page, Boolean includeMenu) in C:\BuildAgent\work\2f923cdb0b81569d\elements\Website.Application\BaseHandlers\XmlOutputHandler.cs:line 274
   at Symplectic.Elements.Website.Application.BaseHandlers.XmlOutputHandler`1.HandleResponseException(Exception doh) in C:\BuildAgent\work\2f923cdb0b81569d\elements\Website.Application\BaseHandlers\XmlOutputHandler.cs:line 517
```

**Comment by jrgriffiniii** [comment]
Reviewing the stack trace, this seems to be somehow related to some issues affecting an XML caching pipeline (raising https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.keynotfoundexception?view=net-6.0). As Elements is not an open source code base, I am going to proceed with contacting Symplectic Support.

**Comment by jrgriffiniii** [comment]
https://support.symplectic.co.uk/support/tickets/359356 has been used to request support with this issue.

**Comment by jrgriffiniii** [comment]
Following the discussion from Symplectic Support, the following may explain why this failure occurred:

> Can you tell us more about what happened after you attempted to upgrade last week? The error you sent is frequently (though not always) caused by the version of the deployed application binaries not being aligned with the version of the database. This can be caused by attempting to manually fix problems with a failed upgrade instead of rolling back and trying again.

**Comment by jrgriffiniii** [comment]
After reviewing the Nagios server alerts for cisdr201w, the following was found:

```
***** Nagios Monitor XI Alert *****

Notification Type: PROBLEM

Service: prod: user: web: https_oaworkflow
Host: cisdr201w.princeton.edu
Address: cisdr201w.princeton.edu
State: CRITICAL

Date/Time: Tue Dec 7 10:46:18 EST 2021

Additional Info:

HTTP CRITICAL: HTTP/1.1 500 Internal Server Error - string CAS Login not found on https://oaworkflow.princeton.edu:443/login.html - 5416 bytes in 0.112 second response time
```

Prior to this message, there were no failures in the production server environment. This confirms my suspicion that this was not immediately triggered by the failed upgrade attempt undertaken on 12/01/2021.

**Comment by jrgriffiniii** [comment]
https://princeton.service-now.com/service?id=ticket&table=incident&sys_id=1c3da87d87d449549cd197183cbb355a&view=sp was created in order to assist with providing additional information to Symplectic support specialists.

**Comment by jrgriffiniii** [comment]
Trying to use the Microsoft Management Console to retrieve more detailed logs yields the following:

![image.png](https://images.zenhubusercontent.com/5b5b6aed981c3204c9ed06c8/6735b2cb-a439-4a1c-9e07-8b006f76eb7b)

**Comment by jrgriffiniii** [comment]
https://princeton.service-now.com/service?id=ticket&table=sc_req_item&sys_id=b7dd34828790cd549cd197183cbb35d9&view=sp is now in progress for requesting the database exports for Symplectic Support.

**Comment by jrgriffiniii** [comment]
I am still struggling to retrieve information regarding the Windows Server using the Event Viewer, as attempting to view error logs simply yields the following:

![image.png](https://images.zenhubusercontent.com/5b5b6aed981c3204c9ed06c8/0fc4e904-24c1-4d60-9082-d7f4d7c272b1)

**Comment by jrgriffiniii** [comment]
Finding a work-around for the Windows Server logs, there is a manual solution for reviewing the activity. For the morning at which the server downtime began, the following is logged:

```
An account was logged off.

Subject:
	Security ID:		PRINCETON\sympapp
	Account Name:		sympapp
	Account Domain:		PRINCETON
	Logon ID:		0x24E59A3F

Logon Type:			5

This event is generated when a logon session is destroyed. It may be positively correlated with a logon event using the Logon ID value. Logon IDs are only unique between reboots on the same computer.
```

The timestamp for this was 12/7/2021 10:21:10 AM.

**Comment by jrgriffiniii** [comment]
This may be unrelated, but it appears that a Puppet Manifest was also run for the `cisdr201w` server prior to this:

![image.png](https://images.zenhubusercontent.com/5b5b6aed981c3204c9ed06c8/0ce5ce09-a12c-483b-b586-3e9f74eeae18)

**Comment by jrgriffiniii** [comment]
After investigating the Puppet log directories and TLS certifications (used for agent/server communication), I could not find any possible cause of these problems.

Further, there exists a directory (`C:\Windows\Logs\FilePurge`) in which the following activity is logged:

```
Processing path C:\Windows\Logs\FilePurge\*.log
Processing path C:\Windows\*.log
ERROR -- C:\Windows\System32\CCM\Cache\* does not exist
ERROR -- C:\Windows\SysWow64\CCM\Cache\* does not exist
Processing path C:\Windows\SoftwareDistribution\Download\*
ERROR -- C:\Windows\$hf_mig$\* does not exist
ERROR -- C:\Windows\$Exch* does not exist
ERROR -- C:\Windows\$NtService* does not exist
ERROR -- C:\Windows\$NtUninst* does not exist
Processing path C:\temp\*
Successfully deleted C:\temp\princeton\Prod Disco 01\Princeton_sn-disc\msvcr100.dll --- [ 1/24/2018 3:44:24 AM ]
Successfully deleted C:\temp\princeton\Prod Disco 01\Princeton_sn-disc\msvcp100.dll --- [ 1/24/2018 3:44:24 AM ]
Successfully deleted C:\temp\princeton\Prod Disco 01\Princeton_sn-disc\get_process_info.exe --- [ 1/24/2018 3:44:32 AM ]
Successfully deleted C:\temp\princeton\Prod Disco 01\Princeton_sn-disc\getEC2Detailsv3.ps1 --- [ 5/18/2020 3:59:07 AM ]
Successfully deleted C:\temp\princeton\Prod Disco 01\Princeton_sn-disc\ConfirmAzureVM.ps1 --- [ 1/24/2018 3:44:24 AM ]
Processing path C:\*\temp\*
Successfully deleted C:\inetpub\temp\IIS Temporary Compressed Files\.NET v4.5\$^_gzip_D^\SYMPLECTIC\ELEMENTS\WEBSITE\JAVASCRIPT\THIRD-PARTY\JQUERY\JQUERY-UI-1.12.1.MIN.JS --- [ 1/29/2019 9:35:39 AM ]
Successfully deleted C:\inetpub\temp\IIS Temporary Compressed Files\.NET v4.5\$^_gzip_D^\SYMPLECTIC\ELEMENTS\WEBSITE\JAVASCRIPT\THIRD-PARTY\JQUERY\JQUERY-3.3.1.MIN.JS --- [ 1/29/2019 9:35:39 AM ]
Successfully deleted C:\inetpub\temp\IIS Temporary Compressed Files\.NET v4.5\$^_gzip_D^\SYMPLECTIC\ELEMENTS\WEBSITE\JAVASCRIPT\THIRD-PARTY\FRAMEWORK-INDEPENDENT\MODERNIZR\MODERNIZR-2.6.2.MIN.JS --- [ 1/29/2019 9:35:38 AM ]
Successfully deleted C:\inetpub\temp\IIS Temporary Compressed Files\.NET v4.5\$^_gzip_D^\SYMPLECTIC\ELEMENTS\WEBSITE\JAVASCRIPT\LIB\DIALOG.JS --- [ 1/29/2019 9:35:36 AM ]
Successfully deleted C:\inetpub\temp\IIS Temporary Compressed Files\.NET v4.5\$^_gzip_D^\SYMPLECTIC\ELEMENTS\WEBSITE\JAVASCRIPT\LIB\COMMON.JS --- [ 1/29/2019 9:35:36 AM ]
Processing path C:\Windows\System32\Logfiles\*
ERROR -- D:\Logfiles\* does not exist
Processing path D:\Logs\Logfiles\*
Successfully deleted D:\Logs\Logfiles\W3SVC1\u_ex211106.log --- [ 11/6/2021 7:58:36 PM ]
```

Might this seem to be indicating that the Internet Information Services (IIS) used to run Elements is having its logs purged daily? I am going to need to proceed with enquiring about FilePurge within OIT.

**Comment by jrgriffiniii** [comment]
This is now awaiting triage with https://princeton.service-now.com/now/workspace/agent/record/incident/b217b58a879401949cd197183cbb3547

**Comment by jrgriffiniii** [comment]
The upload has been completed, and I have informed Symplectic Support of the availability of the database export.

**Comment by jrgriffiniii** [comment]
While Symplectic Support has confirmed that they have received the database export, I have not received an update, and have just now enquired again regarding the status of this request:

![image.png](https://images.zenhubusercontent.com/5b5b6aed981c3204c9ed06c8/f007e04b-75bb-4ee3-af58-2cadcd59276a)

**Comment by jrgriffiniii** [comment]
Symplectic has provided a T-SQL transaction which should restore the database state, and prepare this for another attempted upgrade.

**Comment by jrgriffiniii** [comment]
This should be fully resolved, and should then unblock #556.
